Precompiled headers

Holzhaus opened the pull request:
*Based on #12025.*

No profiling done, I just selected a bunch of headers from the `util` subdirectory plus some other ones that I though might be widely used and don't change that often.

Ideally we'd also add this:

```cmake
target_precompile_headers(mixxx-test REUSE_FROM mixxx-lib)
```

However, this does not work because I think the `libshout-idjc` build messes with the compile options, so the precompiled headers are not reusable and need to be recompiled.

Holzhaus committed:
feat(cmake): Add support for precompiled headers

Holzhaus committed:
fix(configobject): Fix specialization after instantiation errors

Fixes these compiler errors:

    src/preferences/configobject.cpp:333:51: error: specialization of ‘void ConfigObject<ValueType>::setValue(const ConfigKey&, const ResultType&) [with ResultType = QString; ValueType = ConfigValue]’ after instantiation
      333 |         const ConfigKey& key, const QString& value) {
          |                                                   ^
    src/preferences/configobject.cpp:339:48: error: specialization of ‘void ConfigObject<ValueType>::setValue(const ConfigKey&, const ResultType&) [with ResultType = bool; ValueType = ConfigValue]’ after instantiation
      339 |         const ConfigKey& key, const bool& value) {
          |                                                ^
    src/preferences/configobject.cpp:375:58: error: specialization of ‘ResultType ConfigObject<ValueType>::getValue(const ConfigKey&, const ResultType&) const [with ResultType = bool; ValueType = ConfigValue]’ after instantiation
      375 |         const ConfigKey& key, const bool& default_value) const {
          |                                                          ^~~~~
    src/preferences/configobject.cpp:466:61: error: specialization of ‘ResultType ConfigObject<ValueType>::getValue(const ConfigKey&, const ResultType&) const [with ResultType = QString; ValueType = ConfigValue]’ after instantiation
      466 |         const ConfigKey& key, const QString& default_value) const {
          |                                                             ^~~~~

Holzhaus committed:
chore: Remove `moc_included_test.cpp`

This is incompatible with precompiled headers.

Holzhaus committed:
chore(cmake): Only precompile a widely used, stable subset of headers

daschuer commented:
> I think the libshout-idjc build messes with the compile options

Would it be than reasonable to move libshout-idjc to a separate library?

daschuer commented:
Not sure why the https://github.com/mixxxdj/mixxx/pull/12025 commits are still listed.
Can you rebase it to get rid of them?

daschuer commented on CMakeLists.txt:
```diff
@@ -1093,7 +1093,6 @@ add_library(mixxx-lib STATIC EXCLUDE_FROM_ALL
   src/util/logger.cpp
   src/util/logging.cpp
   src/util/mac.cpp
-  src/util/moc_included_test.cpp
```
Why does this no longer work? Does it error with a false positive?

Holzhaus replied:
It always fails on my machine. If you want to try yourself, just checkout this branch and remove the commit to see if it works for you.

Holzhaus commented:
> Not sure why the https://github.com/mixxxdj/mixxx/pull/12025 commits are still listed.
> Can you rebase it to get rid of them?  

Done.

JoergAtGithub commented:
The Windows build failure has it's origin in this name collision:
![grafik](https://github.com/mixxxdj/mixxx/assets/64457745/bc9d2fc7-3a30-4ee8-8e8e-9b32a470d987)

Renaming the variable interface fixed the issue. The build of Mixxx takes 1:50 instead of 6 minutes now. And more than 30 seconds of this time is for libdjinterop.

Holzhaus committed:
fix(hiddevice): Rename `interface` variable due to libshout conflicts

The `libshout-idjc` header file `combaseapi.h` contains the following
line:

    #define interface __STRUCT__

If this file is included (which may happen when using precompiled
headers), the code does not compile on Windows anymore. Is is just a
band-aid - in the long term, inclusion of the `libshout-idjc` should be
reworked if possible.

Holzhaus committed:
fix(cmake): Set project languages properly

JoergAtGithub commented:
Tested various build targets locally all work flawless with a massive improvement in build time.
Full build of x64_off is down to 1'20
Full build of x64_legacy is down to 1'55
Code LGTM and CI passes on all platforms. Thank you!
